commercetools.NET/Common/Client.cs

```diff
@@ -82,22 +83,32 @@ public Client(Configuration configuration)
 
             string url = string.Concat(this.Configuration.ApiUrl, "/", this.Configuration.ProjectKey, endpoint, values.ToQueryString());
 
-            using (HttpClient client = new HttpClient(new ClientLoggingHandler(new HttpClientHandler())))
+            for (int internalServerErrorRetries = -1; internalServerErrorRetries < this.Configuration.InternalServerErrorRetries; internalServerErrorRetries++)
             {
-                client.DefaultRequestHeaders.Accept.Clear();
-                client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
-                client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue(this.Token.TokenType, this.Token.AccessToken);
-                client.DefaultRequestHeaders.UserAgent.ParseAdd(this.UserAgent);
-
-                HttpRequestMessage httpRequestMessage = new HttpRequestMessage(HttpMethod.Get, new Uri(url))
+                using (HttpClient client = new HttpClient(new ClientLoggingHandler(new HttpClientHandler())))
                 {
-                    Version = HttpVersion.Version10
-                };
+                    client.DefaultRequestHeaders.Accept.Clear();
+                    client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+                    client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue(this.Token.TokenType, this.Token.AccessToken);
+                    client.DefaultRequestHeaders.UserAgent.ParseAdd(this.UserAgent);
 
-                HttpResponseMessage httpResponseMessage = await client.SendAsync(httpRequestMessage);
-                response = await GetResponse<T>(httpResponseMessage);
-            }
+                    HttpRequestMessage httpRequestMessage = new HttpRequestMessage(HttpMethod.Get, new Uri(url))
+                    {
+                        Version = HttpVersion.Version10
+                    };
 
+                    HttpResponseMessage httpResponseMessage = await client.SendAsync(httpRequestMessage);
+                    response = await GetResponse<T>(httpResponseMessage);
+                }
+                if (response.StatusCode < 500)
+                {
+                    return response;
+                }
+                else if (this.Configuration.InternalServerErrorRetryInterval > 0)
+                {
+                    Thread.Sleep(this.Configuration.InternalServerErrorRetryInterval);
```

Comment:
Unsure if the Thread.Sleep is correct at this point, as Thread.Sleep stops the whole thread execution. An ```await Task.Delay(this.Configuration.InternalServerErrorRetryInterval);``` would just stop the execution of the task for the specified amount of time and is more inline with the async/await modelling.